#Silver Layer ETL
With the new modeling done, time to ingest the original data into the new tables.

In [0]:
# Imports
import pandas as pd
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, year, month, dayofmonth, weekofyear, monotonically_increasing_id

# Cleaning
Before distributing the data from the bronze layer into the new silver layer tables, we'll check if the data needs cleaning

In [0]:
# For later...
df_bronze = spark.table("bronze.raw_data")

In [0]:
# We will work with data frames for ease of use before converting them to delta tables
df_silver = spark.table("bronze.raw_data")

In [0]:
# Check for duplicates
duplicates = (
    df_silver
    .groupBy(df_silver.columns)
    .count()
    .filter("count > 1")
)
display(duplicates)

# Remove duplicates 
#df_silver = df_silver.dropDuplicates()     

ID Year_Birth Education Marital_Status Income Kidhome Teenhome Dt_Customer Recency MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact Z_Revenue Response count

In [0]:
# Check for NAs
nas = df_silver.select([col(c).isNull().alias(c) for c in df_silver.columns]).groupBy().sum()
display(nas)

# Remove NAs
# Add '#' when no NAs are left
df_silver = df_silver.dropna()

In [0]:
# Total amount of rows should be consistent going forward
display(df_silver.count())

2216

Checking the different columns

In [0]:
display(df_silver.select("Education").distinct())
# Seems fine.

Education Graduation PhD Master Basic 2n Cycle

In [0]:
display(df_silver.select("Marital_Status").distinct())

Marital_Status Single Together Married Divorced Widow Alone Absurd YOLO

Out of the gate, I believe "Absurd" and "YOLO" do not belong to a cleaned dataset, and should be made part of a "Unknown" category for the time being. Meanwhile, "Alone" clearly belongs to the category "Single".

In [0]:
df_silver = df_silver.replace({"Absurd": "Unknown", "YOLO": "Unknown"}, subset=["Marital_Status"])

In [0]:
df_silver = df_silver.replace({"Alone": "Single"}, subset=["Marital_Status"])

In [0]:
display(df_silver.select("Marital_Status").distinct())

Marital_Status Single Together Married Divorced Widow Unknown

# Data Ingestion
From the cleaned dataframe into the new tables

In [0]:
# Branch 2
df_customer = (
    df_silver
    .select(
        "Year_Birth",
        "Education",
        "Marital_Status",
        "Kidhome",
        "Teenhome",
        "Income",
        "Complain"
    )
    .dropDuplicates()   
    )


In [0]:
df_customer.count()

2014

In [0]:
display(df_customer.select("Marital_Status").distinct())

Marital_Status Single Together Married Divorced Widow Unknown

In [0]:
# Checking the schema for conflicts
print(spark.table("bronze.raw_data").printSchema())
print(spark.table("silver.dim_customer").printSchema())

root
 |-- ID: long (nullable = true)
 |-- Year_Birth: long (nullable = true)
 |-- Education: string (nullable = true)
 |-- Marital_Status: string (nullable = true)
 |-- Income: long (nullable = true)
 |-- Kidhome: long (nullable = true)
 |-- Teenhome: long (nullable = true)
 |-- Dt_Customer: date (nullable = true)
 |-- Recency: long (nullable = true)
 |-- MntWines: long (nullable = true)
 |-- MntFruits: long (nullable = true)
 |-- MntMeatProducts: long (nullable = true)
 |-- MntFishProducts: long (nullable = true)
 |-- MntSweetProducts: long (nullable = true)
 |-- MntGoldProds: long (nullable = true)
 |-- NumDealsPurchases: long (nullable = true)
 |-- NumWebPurchases: long (nullable = true)
 |-- NumCatalogPurchases: long (nullable = true)
 |-- NumStorePurchases: long (nullable = true)
 |-- NumWebVisitsMonth: long (nullable = true)
 |-- AcceptedCmp3: long (nullable = true)
 |-- AcceptedCmp4: long (nullable = true)
 |-- AcceptedCmp5: long (nullable = true)
 |-- AcceptedCmp1: long (nullab

In [0]:
# Saving DIM_Customer as a delta table
df_customer.write.format("delta") \
.mode("overwrite") \
.saveAsTable("silver.dim_customer") 

Seems its going to be a large dimensional table, since there are so many permutations of categorical values

### DIM_Promo

In [0]:
df_promo = (
    df_silver
    .select(
        "AcceptedCmp1",
        "AcceptedCmp2",
        "AcceptedCmp3",
        "AcceptedCmp4",
        "AcceptedCmp5",
        "Response"
    )
    .dropDuplicates()
)


In [0]:
# Checking the schema for conflicts
print(spark.table("bronze.raw_data").printSchema())
print(spark.table("silver.dim_promo").printSchema())

root
 |-- ID: long (nullable = true)
 |-- Year_Birth: long (nullable = true)
 |-- Education: string (nullable = true)
 |-- Marital_Status: string (nullable = true)
 |-- Income: long (nullable = true)
 |-- Kidhome: long (nullable = true)
 |-- Teenhome: long (nullable = true)
 |-- Dt_Customer: date (nullable = true)
 |-- Recency: long (nullable = true)
 |-- MntWines: long (nullable = true)
 |-- MntFruits: long (nullable = true)
 |-- MntMeatProducts: long (nullable = true)
 |-- MntFishProducts: long (nullable = true)
 |-- MntSweetProducts: long (nullable = true)
 |-- MntGoldProds: long (nullable = true)
 |-- NumDealsPurchases: long (nullable = true)
 |-- NumWebPurchases: long (nullable = true)
 |-- NumCatalogPurchases: long (nullable = true)
 |-- NumStorePurchases: long (nullable = true)
 |-- NumWebVisitsMonth: long (nullable = true)
 |-- AcceptedCmp3: long (nullable = true)
 |-- AcceptedCmp4: long (nullable = true)
 |-- AcceptedCmp5: long (nullable = true)
 |-- AcceptedCmp1: long (nullab

In [0]:
display(df_promo.count())

35

In [0]:
# Saving DIM_Promo as a delta table
df_promo.write.format("delta") \
.mode("overwrite") \
.saveAsTable("silver.dim_promo") 

### DIM_Calendar

In [0]:
df_date = (
    df_silver
    .select(col("Dt_Customer").alias("full_date"))
    .dropDuplicates()  
    .withColumn("year", year("full_date"))
    .withColumn("month", month("full_date"))
    .withColumn("day", dayofmonth("full_date"))
    .withColumn("week_of_year", weekofyear("full_date"))
)


In [0]:
display(df_date.count())

662

For clariry, this means there are 662 distinct dates in the dataset, allowing us to avoid storing duplicate dates and reducing computational overhead.

In [0]:
# Checking the schema for conflicts
#print(spark.table("bronze.raw_data").printSchema())
print(df_date.printSchema())
print(spark.table("silver.dim_calendar").printSchema())

root
 |-- full_date: date (nullable = true)
 |-- year: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- day: integer (nullable = true)
 |-- week_of_year: integer (nullable = true)

None
root
 |-- id_date: long (nullable = false)
 |-- full_date: date (nullable = true)
 |-- year: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- day: integer (nullable = true)
 |-- week_of_year: integer (nullable = true)

None


In [0]:
# Saving DIM_Calendar as a delta table
df_date.write.format("delta") \
.mode("overwrite") \
.saveAsTable("silver.dim_calendar") 

### FACT_Sales

Only at the end do we ingest the data into the fact table because it will reference all of the other dimensional tables!

In [0]:
%sql
CREATE OR REPLACE TABLE silver.fact_sales
USING DELTA
AS
SELECT
    c.id_customer        AS fk_customer,
    p.id_promo           AS fk_promo,
    d.id_date            AS fk_date,

    b.MntWines,
    b.MntFruits,
    b.MntMeatProducts,
    b.MntFishProducts,
    b.MntSweetProducts,
    b.MntGoldProds,

    b.NumDealsPurchases,
    b.NumWebPurchases,
    b.NumCatalogPurchases,
    b.NumStorePurchases,
    b.NumWebVisitsMonth,
    b.Recency

FROM bronze.raw_data b

LEFT JOIN silver.dim_customer c
    ON  b.Income          = c.Income
    AND b.Kidhome         = c.Kidhome
    AND b.Teenhome        = c.Teenhome
    AND b.Education       = c.Education
    AND b.Marital_Status  = c.Marital_Status
    AND b.Complain        = c.Complain

LEFT JOIN silver.DIM_Promo p
    ON  b.AcceptedCmp1 = p.AcceptedCmp1
    AND b.AcceptedCmp2 = p.AcceptedCmp2
    AND b.AcceptedCmp3 = p.AcceptedCmp3
    AND b.AcceptedCmp4 = p.AcceptedCmp4
    AND b.AcceptedCmp5 = p.AcceptedCmp5
    AND b.Response     = p.Response

LEFT JOIN silver.DIM_Calendar d
    ON b.Dt_Customer = d.full_date;

num_affected_rows num_inserted_rows

In [0]:
display(spark.table("silver.fact_sales").count())

2242

Interesting...
There should not be more than  2216 rows after cleaning the data.

In [0]:
display(spark.table("silver.fact_sales").limit(15))

fk_customer fk_promo fk_date MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth Recency 1750 6 512 635 88 546 172 88 88 3 8 10 4 7 58 1179 9 554 11 1 6 2 1 6 2 1 1 2 5 38 665 9 233 426 49 127 111 21 42 1 8 2 10 4 26 58 9 244 11 4 20 10 3 5 2 2 0 4 6 26 1585 9 229 173 43 118 46 27 15 5 5 3 6 5 94 1288 9 146 520 42 98 0 42 14 2 6 4 10 6 16 1996 9 398 235 65 164 50 49 27 4 7 3 7 6 34 613 9 604 76 10 56 3 1 23 2 4 0 4 8 32 326 6 247 14 0 24 3 3 2 1 3 0 2 9 19 1941 5 548 28 0 6 1 1 13 1 1 0 0 20 68 null 9 607 5 5 6 0 2 1 1 1 0 2 7 11 1994 9 398 6 16 11 11 1 16 1 2 0 3 8 59 1230 9 607 194 61 480 225 112 30 1 3 4 8 2 82 67 9 607 233 2 53 3 5 14 3 6 1 5 6 53 655 9 125 3 14 17 6 1 5 1 1 0 3 8 38

In [0]:
# Check for duplicates in the fact table
duplicates = (
    spark.table("silver.fact_sales")
    .groupBy(spark.table("silver.fact_sales").columns)
    .count()
    .filter("count > 1")
)
display(duplicates)

# Remove duplicates from the fact table
df_fact_sales = spark.table("silver.fact_sales").dropDuplicates()
df_fact_sales.write.format("delta") \
    .mode("overwrite") \
    .saveAsTable("silver.fact_sales")

fk_customer fk_promo fk_date MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth Recency count 217 22 544 1006 22 115 59 68 45 1 7 6 12 3 23 2 1427 9 182 84 5 38 150 12 28 2 4 1 6 7 20 2 79 9 164 384 0 102 21 32 5 3 6 2 9 4 0 2 182 9 558 270 3 27 39 6 99 7 7 1 5 8 69 2 418 16 641 684 100 801 21 66 0 1 6 9 10 2 96 2 1778 9 147 8 4 10 2 2 4 2 3 0 3 5 34 2 235 9 122 112 17 44 34 22 89 1 2 5 3 3 56 2 678 9 260 55 0 6 2 0 4 2 1 1 3 5 99 2 1693 12 16 437 8 206 160 49 42 2 7 10 5 6 53 2 1180 9 358 9 0 6 3 1 3 1 0 0 3 5 4 2 925 9 55 328 9 124 12 24 109 7 4 2 10 3 57 2 1021 19 402 712 26 538 69 13 80 1 3 6 11 1 12 2 198 9 24 64 4 68 7 5 17 4 3 1 4 8 87 2 387 16 445 350 60 568 110 133 145 1 5 6 7 2 70 2 1428 9 147 5 3 11 12 5 9 1 2 0 3 6 30 2 469 8 91 207 26 447 75 0 223 1 5 7 12 2 44 2 939 9 518 135 46 92 65 61 11 2 5 2 7 5 32 2 1171 9 520 1 3 6 10 0 9 1 1 0 3 5 94 2 896 9 449 425 115 292 23 35 79 1 6 6 11 3 46 2 1366 9 518 320 48 133 39 72 48 4 6 2 10 4 65 2 1064 9 178 17 0 14 7 3 2 1 1 0 3 7 83 2 951 9 585 73 3 90 12 1 32 7 4 1 4 8 94 2 312 9 283 15 10 19 8 17 20 3 2 1 3 5 84 2 1963 9 138 620 26 195 34 17 141 1 6 4 13 2 21 2 631 9 562 1 9 7 2 14 10 1 2 0 3 6 80 2 1949 9 367 356 0 107 19 9 117 2 7 1 8 5 34 2 1560 9 538 42 1 18 3 0 4 2 2 0 3 7 13 2 1938 5 341 1 4 10 29 0 36 1 1 1 2 6 52 2 971 9 30 587 43 337 42 87 54 3 11 6 9 6 63 2 113 9 552 54 0 7 0 0 4 1 2 0 3 6 99 2 1245 9 69 297 7 79 0 11 182 2 8 2 4 8 84 2 751 6 366 231 7 137 4 15 39 9 5 1 8 7 30 2 567 9 236 14 1 7 2 1 1 1 2 0 2 9 75 2 270 31 78 1218 16 272 104 0 80 2 5 3 6 6 40 2 644 9 182 864 134 768 150 38 172 1 4 10 5 3 90 2 774 9 348 45 47 52 21 20 25 2 3 1 5 6 92 2 49 9 467 12 4 2 10 6 10 4 2 0 4 5 95 2 1109 12 125 1 2 8 4 3 13 2 1 1 2 7 60 2 1925 9 141 9 0 6 4 1 5 1 1 0 2 9 31 2 1429 5 4 483 84 398 205 84 241 1 3 7 8 1 47 2 1006 9 437 635 15 100 20 7 131 1 9 3 9 6 64 3 1564 9 527 39 0 56 20 8 14 3 3 1 3 7 74 2 669 9 565 9 1 3 0 4 8 1 0 0 3 6 6 2 918 9 161 65 4 16 0 4 11 1 3 0 3 8 43 2 356 9 162 25 0 8 2 0 2 2 2 0 3 7 28 2 860 9 186 44 4 21 6 0 27 2 3 0 3 7 59 2 1390 9 503 3 3 7 6 2 9 1 1 0 2 7 81 2 380 9 292 40 2 15 8 0 6 1 1 1 3 4 46 2 299 12 87 23 1 13 2 2 18 1 1 1 2 7 10 2 1299 9 20 456 19 832 75 118 38 1 4 5 11 3 95 2 68 26 141 1103 0 45 0 0 34 3 4 2 11 8 65 2 160 9 613 378 14 68 19 14 14 2 3 5 8 3 71 2 2001 9 620 15 0 12 7 3 7 3 2 0 3 7 33 2 1016 35 87 656 80 455 52 107 93 2 8 4 8 4 93 2 1775 9 610 269 15 69 15 19 38 2 4 6 4 5 52 2 960 9 137 789 0 142 12 9 38 2 2 4 8 8 63 2 1302 3 582 977 12 253 16 12 101 4 4 3 12 8 45 2 445 9 259 57 0 27 0 0 36 4 3 2 2 8 86 2 854 9 39 10 4 14 4 4 1 1 1 0 3 8 13 2 102 9 336 3 6 5 0 2 6 1 0 0 3 7 74 2 1814 9 494 571 50 142 33 50 159 4 4 5 13 2 78 2 906 9 285 375 152 335 93 91 81 1 5 4 6 2 66 2 1810 9 633 102 9 63 2 9 24 1 4 1 4 6 86 2 1856 9 80 919 0 505 99 30 45 1 4 9 10 2 56 2 537 9 407 19 3 19 3 1 3 2 2 0 3 7 14 2 781 9 192 31 0 7 2 0 6 1 1 0 3 5 49 2 1146 9 402 5 4 13 8 0 4 1 1 0 3 8 56 2 1846 9 330 0 6 3 7 6 12 1 1 0 3 8 99 2 177 9 290 656 38 161 62 47 37 3 8 9 7 7 24 2 357 9 545 398 96 447 220 96 32 1 5 7 6 3 78 2 593 9 147 529 0 356 63 28 0 3 6 7 11 3 63 2 1392 9 320 493 70 324 146 40 84 1 3 7 11 1 77 2 80 9 192 34 4 66 3 10 20 1 2 1 4 7 71 2 462 9 502 227 23 389 42 21 75 1 4 5 7 2 24 2 298 9 531 172 12 112 8 0 15 4 6 1 5 7 70 2 1932 19 155 1478 19 403 0 19 0 1 7 6 6 3 9 2 1982 9 633 30 3 47 19 21 42 1 3 2 2 8 74 2 1497 9 444 7 4 13 15 8 17 2 2 0 4 5 27 2 1136 20 150 1492 38 287 50 57 33 1 5 10 4 6 61 2 205 9 146 30 8 12 8 8 12 1 2 0 4 5 4 2 519 9 323 129 129 259 168 121 83 5 6 3 12 4 13 2 1318 9 661 98 0 106 49 10 106 5 5 1 5 6 26 2 1028 9 182 526 28 135 10 21 99 3 5 3 12 5 83 2 1115 9 85 27 0 5 0 0 3 2 1 0 3 8 94 2 463 8 237 390 22 323 104 35 107 1 6 8 6 3 54 2 486 9 409 1181 107 199 39 30 30 2 7 3 13 8 43 2 499 3 276 1000 19 711 125 115 38 0 4 6 9 1 23 2 1128 9 191 757 80 217 29 80 11 5 9 6 12 6 63 3 994 9 44

Not ideal, but we can clean the dataset once again by removing the 176 duplicates.

In [0]:
display(spark.table("silver.fact_sales").count())

2060